File: packages/sdk-python/evoplatform_sdk/askai.py
"""Client for evo-ai, the Ask AI service. Mirrored with sdk-node's AskAi.

evo-ai is a SEPARATE service from the platform, with its own URL and its own
credentials, so this is a sibling of ``EvoPlatform`` rather than a method on
it.

Two ways to authenticate, matching the two ways an app is built:

- **Service key** — your server holds one ``svc_`` key and names the tenant
  per request. The key authenticates the *application*; ``tenant_id`` states
  which customer's data to search.
- **User access token** — pass a platform-issued token straight through and
  evo-ai reads the tenant from its verified claims.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

import httpx

from .errors import ConfigError, PlatformError

# An LLM composing an answer over retrieved records is not a fast API call —
# the usual client default cuts off answers that were about to succeed.
DEFAULT_TIMEOUT_SECONDS = 120.0


@dataclass
class AskResult:
    answer: str
    #: The question actually executed: with history, evo-ai condenses a
    #: follow-up like "how many?" into a standalone question before searching.
    question: str
    sources: list[dict[str, Any]] = field(default_factory=list)
    action_proposal: dict[str, Any] | None = None
    #: Declined as unrelated to the indexed data, before any AI call.
    #: A refusal is NOT an error — render it differently from a failure.
    gated: bool = False
    #: No usable model for this tenant — an administrator problem.
    unconfigured: bool = False


class AskAi:
    def __init__(
        self,
        url: str,
        *,
        service_key: str | None = None,
        timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS,
        transport: httpx.BaseTransport | None = None,
    ) -> None:
        if not url:
            raise ConfigError("AskAi requires a url")
        self._base = url.rstrip("/")
        self._service_key = service_key
        self._timeout = timeout_seconds
        self._http = httpx.Client(timeout=timeout_seconds, transport=transport)

    def ask(
        self,
        question: str,
        *,
        tenant_id: str | None = None,
        access_token: str | None = None,
        history: list[dict[str, str]] | None = None,
        source_types: list[str] | None = None,
        collection: str = "default",
        allow_actions: bool = False,
    ) -> AskResult:
        """Ask a question against one tenant's indexed data.

        Raises ``PlatformError`` on any non-2xx response or timeout. ``gated``
        and ``unconfigured`` come back on the result, not as exceptions.
        """
        body = {
            "question": question,
            "history": history or [],
            "source_types": source_types,
            "collection": collection,
            "allow_actions": allow_actions,
        }
        res = self._send(
            "/query", body, self._auth_headers(tenant_id, access_token)
        )
        return AskResult(
            answer=res.get("answer") or "",
            question=res.get("question") or question,
            sources=res.get("sources") or [],
            action_proposal=res.get("action_proposal"),
            gated=bool(res.get("gated")),
            unconfigured=bool(res.get("unconfigured")),
        )

    def health(self) -> bool:
        """Liveness check. Cheap enough for a readiness probe."""
        try:
            res = self._http.get(f"{self._base}/health", timeout=5.0)
            return res.status_code == 200
        except httpx.HTTPError:
            return False

    # ---- internals ----

    def _auth_headers(
        self, tenant_id: str | None, access_token: str | None
    ) -> dict[str, str]:
        if access_token:
            # The tenant comes from the token's verified claims. Naming one
            # here too would be ignored, so reject it rather than imply it
            # took effect.
            if tenant_id:
                raise ConfigError(
                    "Pass either access_token or tenant_id, not both — with a "
                    "user token the tenant comes from its verified claims"
                )
            return {"Authorization": f"Bearer {access_token}"}

        if not self._service_key:
            raise ConfigError(
                "AskAi.ask needs either a service_key on the client or an "
                "access_token on the call"
            )
        if not tenant_id:
            # evo-ai answers 400 for this; failing here names the actual mistake.
            raise ConfigError(
                "tenant_id is required when authenticating with a service key "
                "— the key identifies your application, not the customer"
            )
        return {
            "Authorization": f"Bearer {self._service_key}",
            "X-Data-Tenant": str(tenant_id),
        }

    def _send(
        self, path: str, body: Any, headers: dict[str, str]
    ) -> dict[str, Any]:
        try:
            res = self._http.post(
                f"{self._base}{path}", json=body, headers=headers
            )
        except httpx.TimeoutException as exc:
            raise PlatformError(
                504, f"Ask AI request timed out after {self._timeout}s"
            ) from exc
        except httpx.HTTPError as exc:
            raise PlatformError(0, f"Ask AI is unreachable: {exc}") from exc

        try:
            parsed = res.json() if res.text else {}
        except ValueError:
            parsed = res.text
        if res.status_code >= 400:
            detail = (
                str(parsed.get("detail"))
                if isinstance(parsed, dict) and "detail" in parsed
                else f"Ask AI request failed with status {res.status_code}"
            )
            raise PlatformError(res.status_code, detail, parsed)
        return parsed if isinstance(parsed, dict) else {}

File: packages/sdk-node/src/askai.ts
// Evomedia.net EvoPlatform — https://github.com/evomedia-net/evo.platform
// Created by Kelly Michels · dev@evomedia.net
// Licensed under the MIT License. See LICENSE.

import { ConfigError, PlatformError } from './errors';
import { AskAiOptions, AskParams, AskResult } from './types';

/** Default request timeout. An LLM composing an answer over retrieved records
 *  is not a fast API call — the usual 30s client default cuts off answers that
 *  were about to succeed. */
const DEFAULT_TIMEOUT_MS = 120_000;

/**
 * Client for evo-ai, the Ask AI service.
 *
 * evo-ai is a SEPARATE service from the platform, with its own URL and its own
 * credentials, so this is a sibling of {@link EvoPlatform} rather than a method
 * on it.
 *
 * Two ways to authenticate, matching the two ways an app is built:
 *
 * - **Service key** — your server holds one `svc_` key and names the tenant per
 *   request. Use when your app has its own login. The key authenticates the
 *   *application*; `tenantId` states which customer's data to search.
 * - **User access token** — pass a platform-issued token straight through and
 *   evo-ai reads the tenant from its verified claims. Use when your app is in
 *   platform mode.
 *
 * ```ts
 * const ai = new AskAi({ url: process.env.EVOAI_URL!, serviceKey: process.env.EVOAI_SERVICE_KEY });
 * const { answer, sources, gated } = await ai.ask({
 *   question: 'which permits expire this quarter?',
 *   tenantId: session.tenantId,
 *   sourceTypes: ['permit'],
 * });
 * ```
 */
export class AskAi {
  private readonly baseUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchFn: typeof fetch;

  constructor(private readonly opts: AskAiOptions) {
    if (!opts.url) throw new ConfigError('AskAi requires a url');
    this.baseUrl = opts.url.replace(/\/+$/, '');
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchFn = opts.fetchFn ?? fetch;
  }

  /**
   * Ask a question against one tenant's indexed data.
   *
   * Throws {@link PlatformError} on any non-2xx response or timeout — a
   * refusal is NOT an error. Check `gated` (the question was declined as
   * unrelated to the indexed data, before any AI call) and `unconfigured`
   * (no usable model for this tenant, which is an administrator problem)
   * and render those differently from a failure.
   */
  async ask(params: AskParams): Promise<AskResult> {
    const body = {
      question: params.question,
      history: params.history ?? [],
      source_types: params.sourceTypes ?? null,
      collection: params.collection ?? 'default',
      allow_actions: params.allowActions ?? false,
      // Asserts WHICH user is asking, the way X-Data-Tenant asserts which
      // customer. evo-ai keys conversation memory on the pair; a service key
      // is one identity, so omitting this pools every user of a tenant into
      // one memory. Null is the "no user context" case, not an error.
      user_id: params.userId ?? null,
    };

    const res = await this.send('/query', body, this.authHeaders(params));
    return {
      answer: res.answer ?? '',
      // The question actually executed: with history, evo-ai condenses a
      // follow-up like "how many?" into a standalone question before
      // searching. Worth logging, and worth showing when it surprises a user.
      question: res.question ?? params.question,
      sources: res.sources ?? [],
      actionProposal: res.action_proposal ?? null,
      gated: Boolean(res.gated),
      unconfigured: Boolean(res.unconfigured),
    };
  }

  /** Liveness check. Cheap enough for a readiness probe. */
  async health(): Promise<boolean> {
    try {
      const res = await this.fetchFn(`${this.baseUrl}/health`, {
        signal: AbortSignal.timeout(5_000),
      });
      return res.ok;
    } catch {
      return false;
    }
  }

  // ---- internals ----

  private authHeaders(params: AskParams): Record<string, string> {
    if (params.accessToken) {
      // The tenant comes from the token's verified claims. Naming one here
      // too would be ignored, so reject it rather than imply it took effect.
      if (params.tenantId) {
        throw new ConfigError(
          'Pass either accessToken or tenantId, not both — with a user token ' +
            'the tenant comes from its verified claims',
        );
      }
      return { Authorization: `Bearer ${params.accessToken}` };
    }

    if (!this.opts.serviceKey) {
      throw new ConfigError(
        'AskAi.ask needs either a serviceKey on the client or an accessToken on the call',
      );
    }
    if (!params.tenantId) {
      // evo-ai answers 400 for this; failing here names the actual mistake.
      throw new ConfigError(
        'tenantId is required when authenticating with a service key — the key ' +
          'identifies your application, not the customer',
      );
    }
    return {
      Authorization: `Bearer ${this.opts.serviceKey}`,
      'X-Data-Tenant': String(params.tenantId),
    };
  }

  private async send(
    path: string,
    body: unknown,
    headers: Record<string, string>,
  ): Promise<Record<string, any>> {
    let res: Response;
    try {
      res = await this.fetchFn(`${this.baseUrl}${path}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', ...headers },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(this.timeoutMs),
      });
    } catch (err) {
      // A timeout is the most likely failure here, so make it legible rather
      // than surfacing a bare DOMException with no context.
      if (err instanceof Error && (err.name === 'TimeoutError' || err.name === 'AbortError')) {
        throw new PlatformError(504, `Ask AI request timed out after ${this.timeoutMs}ms`);
      }
      throw new PlatformError(
        0,
        `Ask AI is unreachable: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    const text = await res.text();
    let parsed: unknown;
    try {
      parsed = text ? JSON.parse(text) : undefined;
    } catch {
      parsed = text;
    }

    if (!res.ok) {
      const detail =
        typeof parsed === 'object' && parsed !== null && 'detail' in parsed
          ? String((parsed as { detail: unknown }).detail)
          : `Ask AI request failed with status ${res.status}`;
      throw new PlatformError(res.status, detail, parsed);
    }
    return (parsed ?? {}) as Record<string, any>;
  }
}
